pyrules/conditions.py

```diff
@@ -0,0 +1,238 @@
+
+
+class C(object):
+    AND = 'AND'
+    OR = 'OR'
+    XOR = 'XOR'
+    default = AND
+    OPS_MAP = {AND: operator.and_, OR: operator.or_, XOR: operator.xor}
+
+    def __init__(self, _key=None, _value=None, _negate=False, **kwargs):
+        if _key:
+            self._expr = (_key, _value, _negate)
+        else:
+            self._expr = ()
+        for k, v in kwargs.iteritems():
+            self._expr = self.combine(C(k, v))._expr
+
+    def clone(self):
+        """
+        Create a new C-expression that is a copy of this one.
+        """
+        new = self.__class__()
+        new._expr = self._expr
+        return new
+            
+    def combine(self, other, op=default):
+        """
+        Combine another C-expression with this using a logical operator.
+        """
+        assert(isinstance(other, C))
+        self._expr = (
+            op, self._expr, other._expr, False) if self._expr else other._expr
+        return self
+
+    def apply(self, context):
+        """
+        Apply evaluation context to this C-expression. Context should be a dict
+        or dict-like object.
+        """
+        return self._apply_expr(context, self._expr)
+
+    def _apply_expr(self, context, expr):
+        if isinstance(expr[1], tuple):
+            return self.OPS_MAP[expr[0]](
+                self._apply_expr(context, expr[1]),
+                self._apply_expr(context, expr[2]))
+        else:
+            # expr[2] holds negation flag, that's why we xor with it.
+            return ExpressionHandler().evaluate(
+                expr[0], context, expr[1]) ^ expr[2]
+
+    def __repr__(self):
+        return '<C: {}>'.format(self._to_str(self._expr)).encode('utf-8')
+
+    def _to_str(self, val):
+        if not val:
+            return u'N/A'
+        elif isinstance(val[1], tuple):
+            return u'{}({} {} {})'.format(
+                'NOT' if val[-1] else '', self._to_str(val[1]), val[0],
+                self._to_str(val[2]))
+        else:
+            return u'{} {}= {}'.format(
+                val[0], '!' if val[-1] else '=', val[1])            
+
+    def __and__(self, other):
+        return self.clone().combine(other, self.AND)
+
+    def __or__(self, other):
+        return self.clone().combine(other, self.OR)
+
+    def __xor__(self, other):
+        return self.clone().combine(other, self.XOR)
+
+    def __invert__(self):
+        new = self.clone()
+        new._expr = tuple(self._expr[:-1]) + (not self._expr[-1],)
+        return new
+
+    def __eq__(self, other):
+        if not issubclass(other.__class__, C):
+            return False
+        return self._expr == other._expr
+
+
+class ExpressionError(Exception):
+    pass
+
+        
+class ExpressionHandler(object):
+    handlers = [
+        'gt', 'lt', 'gte', 'lte', 'eq', 'neq', 'contains', 'icontains', 'bool']
+
+    def evaluate(self, expression, context, arg):
+        bits = expression.split('__')
```

Comment:
We might have a difference in terminology here. My view is this:

1. from_yaml => parsing => build C expressions / logic evaluator 
2. rule.perfom => apply context => evaluation 

Currently, the C expression built is passed the `key__op,  value` AFAICS in `LogicEvaluator`, and the `apply` on the C expression that results from it happens as part of `evaluate`. All of this happens as a result of a  `TableRule.perform` i.e. at runtime. 

I don't see a benefit in doing all of this at runtime, as the heavy lifting can be done at parse time:

parse time:

1. convert the input format to var__op=value format, and call `to_c_expression` during parsing
2. `var__op` into `key, op` (or bits as it is now), or in other words refactor the first part of `ExpressionHandler.evaluate` such that it can be called at parse time
3. keep the resulting LogicEvaluator instance for later execution

runtime:

4. call evaluator.evaluate(context)

This shouldn't be a big deal - you have implemented all the bits. On the plus side, evaluate becomes a really inexpensive call and multiple executions of the same rules don't need to go through the parsing again and again, as it is now.